src/clang_parser.cpp

```diff
@@ -13,52 +13,135 @@
 
 namespace bpftrace {
 
-std::unordered_map<std::string, CXCursor> indirect_structs;
-std::unordered_set<std::string> unvisited_indirect_structs;
-
 static std::string get_clang_string(CXString string)
 {
   std::string str = clang_getCString(string);
   clang_disposeString(string);
   return str;
 }
 
-static void remove_struct_prefix(std::string &str)
+static void remove_struct_union_prefix(std::string &str)
 {
   if (strncmp(str.c_str(), "struct ", 7) == 0)
     str.erase(0, 7);
+  else if (strncmp(str.c_str(), "union ", 6) == 0)
+    str.erase(0, 6);
 }
 
-static CXCursor get_indirect_field_parent_struct(CXCursor c)
+/*
+ * is_anonymous
+ *
+ * Determine whether the provided cursor points to an anonymous struct.
+ *
+ * This union is anonymous:
+ *   struct { int i; };
+ * This is not, although it is marked as such in LLVM 8:
+ *   struct { int i; } obj;
+ * This is not, and does not actually declare an instance of a struct:
+ *   struct X { int i; };
+ *
+ * The libclang API was changed in LLVM 8 and restored under a different
+ * function in LLVM 9. For LLVM 8 there is no way to properly tell if
+ * a record declaration is anonymous, so we do some hacks here.
+ *
+ * LLVM version differences:
+ *   https://reviews.llvm.org/D54996
+ *   https://reviews.llvm.org/D61232
+ */
+static bool is_anonymous(CXCursor c)
 {
-  CXCursor parent = clang_getCursorSemanticParent(c);
+#if LLVM_VERSION_MAJOR <= 7
+  return clang_Cursor_isAnonymous(c);
+#elif LLVM_VERSION_MAJOR >= 9
+  return clang_Cursor_isAnonymousRecordDecl(c);
+#else // LLVM 8
+  if (!clang_Cursor_isAnonymous(c))
+    return false;
+
+  // In LLVM 8, some structs which the above function says are anonymous
+  // are actually not. We iterate through the siblings of our struct
+  // definition to see if there is a field giving it a name.
+  //
+  // struct Parent                 struct Parent
+  // {                             {
+  //   struct Cursor                 struct Cursor
```

Comment:
This `struct Cursor` comment is a bit misleading - it should just be `struct`. I added the `Cursor` as a way to reference it below, but will change this to be clearer